launchpad/src/launchpad.rs

```diff
@@ -1,9 +1,517 @@
+        let first_ticket_id = self.shuffled_tickets().len() + 1;
+        let last_ticket_id = first_ticket_id + nr_tickets - 1;
+        self.ticket_range_for_address(buyer)
+            .set(&(first_ticket_id, last_ticket_id));
+
+        for i in 0..nr_tickets {
+            self.shuffled_tickets().push(&(first_ticket_id + i));
+        }
+    }
+
+    /// Fisher-Yates algorithm,
+    /// each position is swapped with a random one that's after it.
+    fn shuffle_single_ticket(
+        &self,
+        rng: &mut Random<Self::CryptoApi>,
+        current_ticket_position: usize,
+        last_ticket_position: usize,
+        is_winning_ticket: bool,
+    ) {
+        let rand_index =
+            rng.next_usize_in_range(current_ticket_position + 1, last_ticket_position + 1);
+        self.swap(self.shuffled_tickets(), current_ticket_position, rand_index);
+
+        if is_winning_ticket {
+            let winning_ticket_id = self.shuffled_tickets().get(current_ticket_position);
+            self.set_ticket_status(
+                winning_ticket_id,
+                TicketStatus::Winning {
+                    generation: FIRST_GENERATION,
+                },
+            );
+        }
+    }
+
+    fn swap<T: TopEncode + TopDecode>(
+        &self,
+        mapper: VecMapper<Self::Storage, T>,
+        first_index: usize,
+        second_index: usize,
+    ) {
+        let first_element = mapper.get(first_index);
+        let second_element = mapper.get(second_index);
+
+        mapper.set(first_index, &second_element);
+        mapper.set(second_index, &first_element);
+    }
+
+    fn set_ticket_status(&self, ticket_id: usize, status: TicketStatus) {
+        self.ticket_status().set(ticket_id, &status);
+    }
+
+    fn start_confirmation_period(
+        &self,
+        first_winning_ticket_position: usize,
+        last_winning_ticket_position: usize,
+    ) {
+        let current_epoch = self.blockchain().get_block_epoch();
+        self.confirmation_period_start_epoch().set(&current_epoch);
+
+        let confirmation_period_in_epochs = self.confirmation_period_in_epochs().get();
+        let claim_period_start_epoch = current_epoch + confirmation_period_in_epochs + 1;
+        self.claim_period_start_epoch()
+            .set(&claim_period_start_epoch);
+
+        self.winning_tickets_range()
+            .set(&(first_winning_ticket_position, last_winning_ticket_position));
+        self.current_generation()
+            .update(|current_generation| *current_generation += 1);
+    }
+
+    fn select_new_winners(&self) -> SCResult<()> {
+        let (prev_first_winning_ticket_position, prev_last_winning_ticket_position) =
+            self.winning_tickets_range().get();
+        let winning_tickets =
+            prev_last_winning_ticket_position - prev_first_winning_ticket_position + 1;
+        let redeemed_tickets = self.nr_redeemed_tickets().get();
+        let remaining_tickets = winning_tickets - redeemed_tickets;
+
```

Comment:
I think we can get an early exit if remaining_tickets are 0.
Also we should add a set of checks whether we are in a confirm period / select winners or anything problematic and we could return error and fail the transaction.

Reply:
Added a new functionality known as `LaunchStage`, where a single function (`get_launch_stage`) is doing all the checks to ensure the contract is in the correct state.